gr-blocks/lib/correctiq_auto_impl.cc

```diff
@@ -0,0 +1,179 @@
+                                         float sync_window)
+    : gr::sync_block("correctiq_auto",
+                     gr::io_signature::make(1, 1, sizeof(gr_complex)),
+                     gr::io_signature::make(1, 1, sizeof(gr_complex)))
+{
+    avg_real = 0.0;
+    avg_img = 0.0;
+
+    d_samp_rate = samp_rate;
+    d_freq = freq;
+    d_gain = gain;
+    d_sync_window = sync_window;
+
+    synchronized = false;
+
+    GR_LOG_INFO(d_logger, "Auto Synchronizing...");
+
+    d_max_syncSamples = (long)(d_samp_rate * (double)sync_window);
+
+    message_port_register_in(pmt::mp("rsync"));
+    set_msg_handler(pmt::mp("rsync"),
+                    [this](pmt::pmt_t msg) { this->handle_resync(msg); });
+
+    message_port_register_out(pmt::mp("sync_start"));
+    message_port_register_out(pmt::mp("offsets"));
+}
+
+void correctiq_auto_impl::send_sync_values()
+{
+    pmt::pmt_t meta = pmt::make_dict();
+
+    meta = pmt::dict_add(meta, pmt::mp("real"), pmt::mp(avg_real));
+    meta = pmt::dict_add(meta, pmt::mp("imag"), pmt::mp(avg_img));
+
+    pmt::pmt_t pdu = pmt::cons(meta, pmt::PMT_NIL);
+    message_port_pub(pmt::mp("offsets"), pdu);
+}
+
+void correctiq_auto_impl::send_syncing()
+{
+    pmt::pmt_t pdu = pmt::cons(pmt::intern("syncing"), pmt::from_bool(true));
+
+    message_port_pub(pmt::mp("sync_start"), pdu);
+}
+
+void correctiq_auto_impl::handle_resync(pmt::pmt_t msg)
+{
+    gr::thread::scoped_lock guard(d_mutex);
+
+    GR_LOG_INFO(d_logger, "Auto Synchronizing...");
+    synchronized = false;
+    sync_counter = 0;
+
+    // reset counters on an unsync
+    avg_real = 0.0;
+    avg_img = 0.0;
+
+    send_syncing();
+}
+
+/*
+ * Our virtual destructor.
+ */
+correctiq_auto_impl::~correctiq_auto_impl() {}
+
+double correctiq_auto_impl::get_freq() { return d_freq; }
+
+float correctiq_auto_impl::get_gain() { return d_gain; }
+
+void correctiq_auto_impl::set_freq(double new_value)
+{
+    gr::thread::scoped_lock guard(d_mutex);
+
+    GR_LOG_INFO(d_logger, "Auto Synchronizing...");
+    d_freq = new_value;
+    synchronized = false;
+    sync_counter = 0;
+
+    // reset counters on an unsync
+    avg_real = 0.0;
+    avg_img = 0.0;
+
+    send_syncing();
+}
+
+void correctiq_auto_impl::set_gain(float new_value)
+{
+    gr::thread::scoped_lock guard(d_mutex);
+
+    GR_LOG_INFO(d_logger, "Auto Synchronizing...");
+
+    d_gain = new_value;
+    synchronized = false;
+    sync_counter = 0;
+
+    // reset counters on an unsync
+    avg_real = 0.0;
+    avg_img = 0.0;
+
+    send_syncing();
+}
+
+int correctiq_auto_impl::work(int noutput_items,
+                              gr_vector_const_void_star& input_items,
+                              gr_vector_void_star& output_items)
+{
+    gr::thread::scoped_lock guard(d_mutex);
+
+    const gr_complex* in = (const gr_complex*)input_items[0];
+    gr_complex* out = (gr_complex*)output_items[0];
+
+    for (int i = 0; i < noutput_items; i++) {
```

Comment:
the overall computation is just "out = in - const(avg_real, avg_imag);", regardless of whether synched or not. When not in sync, the "const" is changing, and hence the loop is required. But, when in sync, seems like you could create a Volk vectors and then use `volk_32fc_x2_add_32fc` to add them together. You'd have to keep track of the Volk vectors to make sure they are long enough, which does add a touch of complexity, but I tend to believe it would be a lot faster than doing this item by item. In the constructor you could use `set_max_noutput_items` to some reasonable amount and allocate the Volk buffers to that value (in items). Just thinking of alternative optimizations to this algorithm that wouldn't be too difficult to implement.